graylog2-web-interface/src/views/logic/views/UseCreateViewForEvent.ts

```diff
@@ -0,0 +1,200 @@
+  const histogram = resultHistogram();
+  const messageTable = allMessagesTable(undefined, streamDecorators);
+  const summaryAggregations = {
+    fnSeries: [],
+    rowPivots: [],
+    title: 'Summary: ',
+  };
+  const needsSummaryAggregations = aggregations.length > 1;
+  const SUMMARY_ROW_DELTA = needsSummaryAggregations ? AGGREGATION_WIDGET_HEIGHT : 0;
+  const { aggregationWidgets, aggregationTitles, aggregationPositions } = aggregations.reduce((res, { field, value, expr, fnSeries }, index) => {
+    const rowPivots = [pivotForField(uniq([field, ...groupBy]), new FieldType('value', [], []))];
+    const fnSeriesForFunc = Series.forFunction(fnSeries);
+    const direction = ['>', '>=', '=='].includes(expr) ? Direction.Descending : Direction.Ascending;
+    const sort = [new SortConfig(SortConfig.SERIES_TYPE, fnSeries, direction)];
+    const widget = getAggregationWidget({ rowPivots, fnSeries: [fnSeriesForFunc], sort });
+    const widgetId = widget.id;
+    const title = `${fnSeries} ${expr} ${value}`;
+    const isEven = (index + 1) % 2 === 0;
+    const col = isEven ? 7 : 1;
+    const HEIGHT_DELTA = index >= 2 ? AGGREGATION_WIDGET_HEIGHT : 0;
+    const row = Math.ceil((index + 1) / 2) + HEIGHT_DELTA + SUMMARY_ROW_DELTA;
+    const position = new WidgetPosition(col, row, AGGREGATION_WIDGET_HEIGHT, 6);
+    res.aggregationWidgets.push(widget);
+    res.aggregationTitles[widgetId] = title;
+    res.aggregationPositions[widgetId] = position;
+
+    if (needsSummaryAggregations) {
+      summaryAggregations.fnSeries.push(fnSeries);
+      summaryAggregations.rowPivots.push(field);
+      summaryAggregations.title = `${summaryAggregations.title} ${title}`;
+    }
+
+    return res;
+  }, { aggregationTitles: {}, aggregationWidgets: [], aggregationPositions: {} });
+
+  const widgets = [
+    ...aggregationWidgets,
+    histogram,
+    messageTable,
+  ];
+
+  const titles = {
+    widget: {
+      ...aggregationTitles,
+      [histogram.id]: 'Message Count',
+      [messageTable.id]: 'All Messages',
+    },
+  };
+
+  const positions = {
+    ...aggregationPositions,
+    [histogram.id]: new WidgetPosition(1, AGGREGATION_WIDGET_HEIGHT * aggregationWidgets.length + 1 + SUMMARY_ROW_DELTA, 2, Infinity),
+    [messageTable.id]: new WidgetPosition(1, AGGREGATION_WIDGET_HEIGHT * aggregationWidgets.length + 3 + SUMMARY_ROW_DELTA, 6, Infinity),
+  };
+
+  if (needsSummaryAggregations) {
+    const summaryAggregationWidget = getAggregationWidget({
+      rowPivots: [pivotForField(uniq([...summaryAggregations.rowPivots, ...groupBy]), new FieldType('value', [], []))],
+      fnSeries: summaryAggregations.fnSeries.map((s) => Series.forFunction(s)),
+    });
+    widgets.push(summaryAggregationWidget);
+    titles.widget[summaryAggregationWidget.id] = summaryAggregations.title;
+    positions[summaryAggregationWidget.id] = new WidgetPosition(1, 1, AGGREGATION_WIDGET_HEIGHT, Infinity);
+  }
+
+  return { titles, widgets, positions };
```

Comment:
This function is a little bit hard to read, maybe you can declutter it by separating the concerns and moving code blocks into dedicated functions?